Case search screen subtitle/description text

nospame opened the pull request:
## Product Description
Adds "Search Screen Subtitle" to case search screen options and shows this text on case search query screen.

In app manager:
<img width="500" alt="Screen Shot 2022-11-16 at 3 20 45 PM" src="https://user-images.githubusercontent.com/100609770/202315849-cfa43feb-8aae-499c-a450-53a407b6e407.png">


In case search:
![Screen Shot 2022-11-14 at 10 08 27 AM](https://user-images.githubusercontent.com/100609770/201734329-a222f36b-61ba-4312-bafe-a67b50abe4fa.png)


## Technical Summary
[USH-2366](https://dimagi-dev.atlassian.net/browse/USH-2366)

- Adds `description` to app manager screen with a new `textarea_trans` field in the style of `input_trans`. 
- Saves to CaseSearch model as a LabelProperty. 
- Uses app_strings and id_strings to make text translatable, with additions to bulk upload/download modules to allow bulk translation.
- Sends `<description>` field as part of RemoteRequestQuery XML to be read by Formplayer. 
- Renders with markdown on case search screen using `markdown-it` and `DOMPurify`.

Related changes to commcare-core in https://github.com/dimagi/commcare-core/pull/1164 and formplayer in https://github.com/dimagi/formplayer/pull/1267

## Feature Flag
SYNC_SEARCH_CASE_CLAIM

## Safety Assurance

### Safety story
Tested workflows locally including translations, as well as using "dirty" strings to check for proper sanitizing. Ran automated tests known to touch these areas and added description text to each.

### Automated test coverage
Added coverage in case claim bulk translation tests.

### QA Plan
Will send to QA to ensure proper entry, translation, and display of description text.
[QA-4676](https://dimagi-dev.atlassian.net/browse/QA-4676?atlOrigin=eyJpIjoiODljZmQ2NWM1NzhlNDIwNjg4ODQ5MmMxYjc2NTBmZjYiLCJwIjoiaiJ9)

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Add case search description text to app manager

nospame committed:
Add description text to CaseSearch model

nospame committed:
Add app strings + id strings for description text

nospame committed:
Add description to translation upload / download modules

nospame committed:
Add description to RemoteRequestQuery and build_remote_request_queries

nospame committed:
Render description on case search screen

nospame committed:
Use textarea for description input

nospame committed:
Add query screen description to tests + css style for links

nospame committed:
Stickler whitespace

nospame committed:
Add description block to xml tests

Robert-Costello commented on corehq/apps/cloudcare/templates/formplayer/query_view.html:
```diff
@@ -3,7 +3,10 @@
 
 <script type="text/template" id="query-view-list-template">
   <form>
-    <p><h2 tabindex="0"><%- title %></h2></p>
```
Not clear to me what the `<p>` tag was giving us here, so I'm curious about your reasons for removing it (and for using a `<p>` for `description`). Did it have a noticeable affect?

nospame replied:
It seemed like the intent was maybe to ensure a line break on either side of the heading (which displays on its own line by default), but the actual rendered html put an entire new paragraph on either side of the heading to try to 'fix' the incorrect nesting. So it would come out as:
```html
<p></p>
<h2>(title)</h2>
<p></p>
```

Robert-Costello commented on corehq/apps/app_manager/templatetags/xforms_extras.py:
```diff
@@ -140,6 +140,30 @@ def input_trans(name, langs=None, input_name='name', input_id=None, data_bind=No
+        "data_bind_attribute": data_bind_attribute
+    })
+
+    print(options)
```
This should go away, yeah?

nospame replied:
Agh yep, thank you for catching.

orangejenny commented on corehq/apps/app_manager/templatetags/xforms_extras.py:
```diff
@@ -140,6 +140,30 @@ def input_trans(name, langs=None, input_name='name', input_id=None, data_bind=No
 
 
+@register.simple_tag
+def textarea_trans(name, langs=None, input_name='name', input_id=None, data_bind=None):
```
There's a fair amount of duplication with `input_trans`. Thoughts on instead making the node name a param? (Which admittedly makes the function name make less sense.)

Robert-Costello replied:
> Thoughts on instead making the element name a param? 

You'd need to update other places where it's used by pass in the new element name param, "input" in current usages, and also account for `type` of input, which isn't a consideration for `text area`.

orangejenny replied:
It could default to `"input"`, but yeah, fair point that `type` isn't relevant for textareas.

nospame replied:
I have conflicting thoughts - I think it makes sense to combine from the side of them being text entries with the same basic functionality, and almost did so. At the same time, when I think about the `<input>` element, it's used for so many things that aren't just `input type="text"` that I would hesitate to include the `<textarea>` element in the same function.

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -2260,6 +2260,7 @@ class CaseSearch(DocumentSchema):
     data_registry_workflow = StringProperty(exclude_if_none=True)  # one of REGISTRY_WORKFLOW_*
     additional_registry_cases = StringListProperty()               # list of xpath expressions
     title_label = LabelProperty(default={})
+    description_text = LabelProperty(default={})
```
Naming nitpick: Why not just `description`? "description" implies that it's text, and shorter variable names are generally better.

orangejenny replied:
Also on this topic: I've grown draconian about being as consistent as possible with naming, so in this PR, I'd use the same name for the app manager model (CaseSearch) and the xml model (RemoteRequestQuery), unless there's a reason not to.

nospame replied:
Makes sense to me, definitely happy to switch to `description`. I think I made it two words following the example of `title_label` without thinking too hard.

orangejenny commented on corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:
```diff
@@ -180,6 +180,11 @@ def build_title(self):
         )
 
+    def build_description(self):
+        return Display(
```
It's late for this feedback, but it also would be fine to only add a description node if there is a description, which should save you from updating all the text XML files.

nospame replied:
That makes sense, probably cleaner to not include it where it would be empty. And not a ton of work to go back and change. Does something like this as a sub for [line 205](https://github.com/dimagi/commcare-hq/blob/eb1efaddfd7a763768e7b0400527437f589602ff/corehq/apps/app_manager/suite_xml/post_process/remote_requests.py#L205) seem to make sense for it?
```python
description=self.build_description() if self.module.search_config.description != {} else None,
```

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_properties.html:
```diff
@@ -182,6 +182,16 @@ <h4 class="panel-title panel-title-nolink">{% trans "Search and Claim Options" %
             </div>
           </div>
+          <div class="form-group">
+            <label for="search_screen_subtitle" class="{% css_label_class %} control-label">
```
The value for the `for` attribute should be the id of the input, so currently `description_text`

orangejenny commented:
> Renders on case search query screen using DOMPurify to allow html markup.

Could this support markdown instead of HTML? App builders are a bit more familiar with that - there's a markdown option for case detail properties, and markdown is used in form builder as well.

nospame commented:
> Could this support markdown instead of HTML? App builders are a bit more familiar with that - there's a markdown option for case detail properties, and markdown is used in form builder as well.

@orangejenny I think this was me getting mark _down_ confused with html mark _up_. Looks like it's pretty straightforward to have it render markdown on the query screen. Should it get a preview box in app manager the same way a question's Display Text does, for example?

nospame committed:
Use 'description' everywhere

nospame committed:
Render description with markdown

orangejenny commented:
> Should it get a preview box in app manager the same way form builder does, for example?

It'd be nice, but it's also fine to call it out of scope.

nospame committed:
Update description help text

nospame committed:
Only generate description node if needed + update test xml

nospame committed:
Fix css

Previously changed link color on other pages, updated to only affect case search description

nospame committed:
Consolidate textarea_trans into input_trans

Add element_type option to input_trans defaulting to 'input_text', add test coverage

nospame committed:
Specify id to bind query properties to

Fixes QA-4680 where properties would be duplicated if there were multiple tables
